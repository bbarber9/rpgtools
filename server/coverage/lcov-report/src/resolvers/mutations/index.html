
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/resolvers/mutations</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> src/resolvers/mutations</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">4.43% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>20/451</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">2.2% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>4/182</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">1.52% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>1/66</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">4.51% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>20/443</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="authentication-mutations.js"><a href="authentication-mutations.js.html">authentication-mutations.js</a></td>
	<td data-value="33.33" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 33%"></div><div class="cover-empty" style="width: 67%"></div></div>
	</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="30" class="abs low">10/30</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="10" class="abs low">4/10</td>
	<td data-value="20" class="pct low">20%</td>
	<td data-value="5" class="abs low">1/5</td>
	<td data-value="34.48" class="pct low">34.48%</td>
	<td data-value="29" class="abs low">10/29</td>
	</tr>

<tr>
	<td class="file low" data-value="authorization-mutations.js"><a href="authorization-mutations.js.html">authorization-mutations.js</a></td>
	<td data-value="0.92" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0.92" class="pct low">0.92%</td>
	<td data-value="109" class="abs low">1/109</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="56" class="abs low">0/56</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="12" class="abs low">0/12</td>
	<td data-value="0.95" class="pct low">0.95%</td>
	<td data-value="105" class="abs low">1/105</td>
	</tr>

<tr>
	<td class="file low" data-value="folder-mutations.js"><a href="folder-mutations.js.html">folder-mutations.js</a></td>
	<td data-value="5.71" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 5%"></div><div class="cover-empty" style="width: 95%"></div></div>
	</td>
	<td data-value="5.71" class="pct low">5.71%</td>
	<td data-value="35" class="abs low">2/35</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="16" class="abs low">0/16</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="5.71" class="pct low">5.71%</td>
	<td data-value="35" class="abs low">2/35</td>
	</tr>

<tr>
	<td class="file low" data-value="image-mutations.js"><a href="image-mutations.js.html">image-mutations.js</a></td>
	<td data-value="2.08" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 2%"></div><div class="cover-empty" style="width: 98%"></div></div>
	</td>
	<td data-value="2.08" class="pct low">2.08%</td>
	<td data-value="96" class="abs low">2/96</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="28" class="abs low">0/28</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="27" class="abs low">0/27</td>
	<td data-value="2.08" class="pct low">2.08%</td>
	<td data-value="96" class="abs low">2/96</td>
	</tr>

<tr>
	<td class="file low" data-value="server-mutations.js"><a href="server-mutations.js.html">server-mutations.js</a></td>
	<td data-value="4.55" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 4%"></div><div class="cover-empty" style="width: 96%"></div></div>
	</td>
	<td data-value="4.55" class="pct low">4.55%</td>
	<td data-value="22" class="abs low">1/22</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="5" class="pct low">5%</td>
	<td data-value="20" class="abs low">1/20</td>
	</tr>

<tr>
	<td class="file low" data-value="user-mutations.js"><a href="user-mutations.js.html">user-mutations.js</a></td>
	<td data-value="20" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 20%"></div><div class="cover-empty" style="width: 80%"></div></div>
	</td>
	<td data-value="20" class="pct low">20%</td>
	<td data-value="5" class="abs low">1/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="20" class="pct low">20%</td>
	<td data-value="5" class="abs low">1/5</td>
	</tr>

<tr>
	<td class="file low" data-value="wiki-mutations.js"><a href="wiki-mutations.js.html">wiki-mutations.js</a></td>
	<td data-value="1.59" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 1%"></div><div class="cover-empty" style="width: 99%"></div></div>
	</td>
	<td data-value="1.59" class="pct low">1.59%</td>
	<td data-value="63" class="abs low">1/63</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="30" class="abs low">0/30</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	<td data-value="1.61" class="pct low">1.61%</td>
	<td data-value="62" class="abs low">1/62</td>
	</tr>

<tr>
	<td class="file low" data-value="wiki-page-mutations.js"><a href="wiki-page-mutations.js.html">wiki-page-mutations.js</a></td>
	<td data-value="10" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 10%"></div><div class="cover-empty" style="width: 90%"></div></div>
	</td>
	<td data-value="10" class="pct low">10%</td>
	<td data-value="10" class="abs low">1/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="10" class="pct low">10%</td>
	<td data-value="10" class="abs low">1/10</td>
	</tr>

<tr>
	<td class="file low" data-value="world-mutations.js"><a href="world-mutations.js.html">world-mutations.js</a></td>
	<td data-value="1.23" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 1%"></div><div class="cover-empty" style="width: 99%"></div></div>
	</td>
	<td data-value="1.23" class="pct low">1.23%</td>
	<td data-value="81" class="abs low">1/81</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="30" class="abs low">0/30</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="1.23" class="pct low">1.23%</td>
	<td data-value="81" class="abs low">1/81</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Tue Mar 17 2020 14:01:18 GMT-0600 (Mountain Daylight Time)
            </div>
        </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    