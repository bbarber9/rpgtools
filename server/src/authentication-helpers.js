import jwt from "jsonwebtoken";
import {v4 as uuidv4} from "uuid";
import {User} from "./models/user";
import {ANON_USERNAME} from "../../common/src/permission-constants";

export const authenticated = next => (root, args, context, info) => {
	if (!context.currentUser || context.currentUser.username === ANON_USERNAME) {
		throw new Error(`Authentication is required for this action`);
	}

	return next(root, args, context, info);
};

export const createTokens = async user => {
	const accessToken = jwt.sign({userId: user._id}, process.env['ACCESS_TOKEN_SECRET'], {expiresIn: ACCESS_TOKEN_MAX_AGE.string});
	let version = uuidv4();
	const refreshToken = jwt.sign({version: version, userId: user._id}, process.env['REFRESH_TOKEN_SECRET'], {expiresIn: REFRESH_TOKEN_MAX_AGE.string});
	user.tokenVersion = version;
	await user.save();
	return {accessToken, refreshToken};
};

const getCurrentUser = async (userId) => {
	return User.findById(userId)
		.populate({path: 'currentWorld', populate: {path: 'wikiPage'}})
		.populate({path: 'permissions', populate: 'subject'})
		.populate({path: 'roles', populate: 'permissions'});
};

export const createSessionContext = async ({req, res}) => {

	let currentUser = null;

	const refreshToken = req.cookies["refreshToken"];
	const accessToken = req.cookies["accessToken"];
	if (refreshToken || accessToken) {
		try {
			let data = jwt.verify(accessToken, process.env['ACCESS_TOKEN_SECRET'], {maxAge: ACCESS_TOKEN_MAX_AGE.string});
			currentUser = await getCurrentUser(data.userId);
		} catch (e) {
			// accessToken is expired
			console.log(`Access token expired because ${e.message}`);
			res.clearCookie('accessToken');
			try {
				let data = jwt.verify(refreshToken, process.env['REFRESH_TOKEN_SECRET'], {maxAge: REFRESH_TOKEN_MAX_AGE.string});
				currentUser = await getCurrentUser(data.userId);
				// if refreshToken is still valid issue new access token and refresh token
				if (currentUser && currentUser.tokenVersion === data.version) {
					console.log('Access token and refresh token renewed b/c refresh token still valid');
					let tokens = await createTokens(currentUser);
					res.cookie('accessToken', tokens.accessToken, {maxAge: ACCESS_TOKEN_MAX_AGE.ms});
					res.cookie('refreshToken', tokens.refreshToken, {maxAge: REFRESH_TOKEN_MAX_AGE.ms});
				} else {
					// refreshToken was invalidated
					console.log(`Refresh token expired because version did not match ${currentUser.tokenVersion} !== ${data.version}`);
					currentUser = null;
					res.clearCookie('refreshToken');
				}
			} catch (e){
				// refreshToken is expired
				console.log(`Refresh token expired because ${e.message}`);
			}
		}
	}

	if(!currentUser){
		currentUser = new User({username: ANON_USERNAME});
	}

	await currentUser.recalculateAllPermissions();

	return {
		res,
		currentUser
	};
};

export const ACCESS_TOKEN_MAX_AGE = {string: '15min', ms: 1000 * 60 * 15};
export const REFRESH_TOKEN_MAX_AGE = {string: '1d', ms: 1000 * 60 * 60 * 24};