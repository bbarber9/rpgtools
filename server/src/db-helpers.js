import {PermissionAssignment} from "./models/permission-assignement";
import {Role} from './models/role';
import {User} from './models/user';
import {GridFSBucket} from "mongodb";
import mongoose from "mongoose";

export const cleanUpPermissions = async (subjectId) => {
	const assignments = await PermissionAssignment.find({subject: subjectId});
	for(let assignment of assignments){
		const roles = await Role.find({permissions: assignment._id});
		for(let role of roles){
			role.permissions = role.permissions.filter((permission) => {return !permission._id.equals(assignment._id);});
			await role.save();
		}
		const users = await User.find({permissions: assignment._id});
		for(let user of users){
			user.permissions = user.permissions.filter((permission) => {return !permission._id.equals(assignment._id);});
			await user.save();
		}
		await PermissionAssignment.deleteOne({_id: assignment._id});
	}
};

export const createGfsFile = async (filename, readStream) => {
	return new Promise((resolve, reject) => {

		const gfs = new GridFSBucket(mongoose.connection.db);
		const writeStream = gfs.openUploadStream(filename);

		writeStream.on('finish', (file) => {
			resolve(file._id);
		});

		writeStream.on('error', (err) => {
			reject(err);
			throw err;
		});

		readStream.pipe(writeStream);
	});
};