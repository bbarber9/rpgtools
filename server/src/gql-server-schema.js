import gql from "graphql-tag";

export const typeDefs = gql`

    type Query {
        currentUser: User
        
        serverConfig: ServerConfig!
        
        """Get a world by id"""
        world(worldId: ID): World
        """Get all worlds"""
        worlds(page: Int): WorldPaginatedResult
        
        """Get a wiki by id"""
        wiki(wikiId: ID!): WikiPage
        """Search for a wiki page by name"""
        wikis(worldId: ID!, name: String!, type: String): WikiPagePaginatedResult!
        
        """Search for users by username"""
        users(username: String!): UserPaginatedResult!
        
        """Search for roles by name"""
        roles(worldId: ID!, name: String!): RolePaginatedResult!
        
        """Get a game by id"""
        game(gameId: ID!): Game!
        
        """Get all games that the current user is a part of"""
        myGames: [Game!]!
        
        """Get all models for a world"""
        models(worldId: ID!): [Model!]!
        
        """Get permissions for the current authenticated user for a particular world"""
        myPermissions(worldId: ID!): [PermissionAssignment!]!
    }
  
    type Mutation{
        unlockServer(unlockCode: String!, email: String!, username: String!, password: String!): Boolean
        generateRegisterCodes(amount: Int!): ServerConfig!
        
        login(username: String!, password:  String!): User!
        logout: String!
        register(registerCode: String!, email: String!, username: String!, password: String!): User!
        setCurrentWorld(worldId: ID!): User!
        
        createWorld(name: String!, public: Boolean!): World!
        renameWorld(worldId: ID!, newName: String!): World!
        
        createRole(worldId: ID!, name: String!): World!
        deleteRole(roleId: ID!): World!
        addUserRole(userId: ID!, roleId: ID!): World!
        removeUserRole(userId: ID!, roleId: ID!): World!
        
        """ Grant a permission to a user """
        grantUserPermission(userId: ID!, permission: String!, subjectId: ID!, subjectType: String!): PermissionControlled!
        """ Revoke a assigned permission from a user """
        revokeUserPermission(userId: ID!, permissionAssignmentId: ID!): PermissionControlled!
        """ Grant a permission to a role """
        grantRolePermission(roleId: ID!, permission: String!, subjectId: ID!, subjectType: String!): Role!
        """ Revoke a permission from a role """
        revokeRolePermission(roleId: ID!, permissionAssignmentId: ID!): Role!
        
        createFolder(name: String!, parentFolderId: ID!): World!
        renameFolder(folderId: ID!, name: String!): WikiFolder!
        deleteFolder(folderId: ID!): World!
        moveFolder(folderId: ID!, parentFolderId: ID!): World!
        
        """ Creates a generic wiki page """
        createWiki(name: String!, folderId: ID!): World!
        """ Deletes any wiki page of any type """
        deleteWiki(wikiId: ID!): World!
        """ Updates any wiki page of any type """
        updateWiki(wikiId: ID!, name: String, content: Upload, coverImageId: ID, type: String): WikiPage!
        """ Update place specific attributes """
        updatePlace(placeId: ID!, mapImageId: ID, pixelsPerFoot: Int): Place!
        
        createImage(file: Upload!, worldId: ID!, chunkify: Boolean): Image!
        
        createPin(mapId: ID!, x: Float!, y: Float!, wikiId: ID): World!
        updatePin(pinId: ID!, pageId: ID): World!
        deletePin(pinId: ID!): World!
        
        createGame(worldId: ID!, password: String): Game!
        joinGame(gameId: ID!, password: String): Game!
        leaveGame(gameId: ID!): Boolean!
        gameChat(gameId: ID!, message: String!): Game!
        setGameMap(gameId: ID!, placeId: ID!): Game!
        addStroke(gameId: ID!, path: [PathNodeInput!]!, type: String!, size: Int!, color: String!, fill: Boolean!, strokeId: ID!): Game!
        addModel(gameId: ID!, modelId: ID!): Game!
        setModelPosition(gameId: ID!, positionedModelId: ID!, x: Float!, z: Float, rotation: Float!): PositionedModel!
        
        createModel(name: String!, file: Upload!, worldId: ID!, depth: Float!, width: Float!, height: Float!): Model!
        updateModel(modelId: ID!, name: String!, file: Upload, depth: Float!, width: Float!, height: Float!): Model!
        deleteModel(modelId: ID!): Model!
    }
  
    type Subscription {
		gameChat(gameId: ID!): GameMessage!
		gameRosterChange(gameId: ID!): Game!
		gameMapChange(gameId: ID!): Game!
		gameStrokeAdded(gameId: ID!): Stroke!
		gameModelAdded(gameId: ID!): PositionedModel!
		gameModelPositioned(gameId: ID!): PositionedModel!
	}
  
    type User {
        _id: ID!
        username: String!
        email: String
        currentWorld: World
        roles: [Role!]!
        permissions: [PermissionAssignment!]!
    }
  
    interface PermissionControlled {
        accessControlList: [PermissionAssignment!]!
        canWrite: Boolean!
        canAdmin: Boolean!
        _id: ID!
    }
  
	type World implements PermissionControlled {
		_id: ID!
		name: String!
		wikiPage: Place
		rootFolder: WikiFolder
		roles: [Role!]!
		pins: [Pin!]!
		folders: [WikiFolder!]!
		accessControlList: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
		canAddRoles: Boolean!
		canHostGame: Boolean!
		canAddModels: Boolean!
		currentUserPermissions: [PermissionAssignment!]!
	}
	
	type WorldPaginatedResult {
        docs: [World!]!
        totalDocs: Int!
        limit: Int!
        page: Int!
        totalPages: Int!
        pagingCounter: Int!
        hasPrevPage: Boolean!
        hasNextPage: Boolean!
        prevPage: Int
        nextPage: Int
    }
    
    type UserPaginatedResult {
        docs: [User!]!
        totalDocs: Int!
        limit: Int!
        page: Int!
        totalPages: Int!
        pagingCounter: Int!
        hasPrevPage: Boolean!
        hasNextPage: Boolean!
        prevPage: Int
        nextPage: Int
    }
    
    type RolePaginatedResult {
        docs: [Role!]!
        totalDocs: Int!
        limit: Int!
        page: Int!
        totalPages: Int!
        pagingCounter: Int!
        hasPrevPage: Boolean!
        hasNextPage: Boolean!
        prevPage: Int
        nextPage: Int
    }
    
    type WikiPagePaginatedResult {
        docs: [Role!]!
        totalDocs: Int!
        limit: Int!
        page: Int!
        totalPages: Int!
        pagingCounter: Int!
        hasPrevPage: Boolean!
        hasNextPage: Boolean!
        prevPage: Int
        nextPage: Int
    }
	
	interface WikiPage {
		_id: ID!
		name: String!
		content: String
		world: World!
		coverImage: Image
		type: String!
	}
	
	type Article implements WikiPage & PermissionControlled {
		_id: ID!
		name: String!
		content: String
		world: World!
		coverImage: Image
		type: String!
		accessControlList: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
	}
	
	type Place implements WikiPage & PermissionControlled {
		_id: ID!
		name: String!
		content: String
		world: World!
		coverImage: Image
		mapImage: Image
		pixelsPerFoot: Int
		type: String!
		accessControlList: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
	}
	
	type Person implements WikiPage & PermissionControlled {
		_id: ID!
		name: String!
		content: String
		world: World!
		coverImage: Image
		type: String!
		accessControlList: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
	}
	
	type WikiFolder implements PermissionControlled{
		_id: ID!
		name: String!
		world: World!
		pages: [WikiPage!]!
		children: [WikiFolder!]!
		accessControlList: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
	}
	
	type Image {
		_id: ID!
		world: World!
		height: Int!
		width: Int!
		chunkHeight: Int!
		chunkWidth: Int!
		chunks: [Chunk!]!
		icon: Image!
		name: String!
	}
	
	type Chunk {
		_id: ID!
		image: Image!
		x: Int!
		y: Int!
		width: Int!
		height: Int!
		fileId: ID!
	}
	
	type Role implements PermissionControlled{
		_id: ID!
		name: String!
		accessControlList: [PermissionAssignment!]!
		members: [User!]!
		world: World
		permissions: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
	}
	
	type PermissionAssignment {
		_id: ID!
		permission: String!
		subject: PermissionControlled!
		subjectType: String!
		canWrite: Boolean!
		users: [User!]!
		roles: [Role!]!
	}
	
	type Pin {
		_id: ID!
		map: Place!
		x: Float!
		y: Float!
		page: WikiPage
		canWrite: Boolean!
	}
	
	type ServerConfig implements PermissionControlled{
		_id: ID!
		version: String!
		registerCodes: [String!]!
		accessControlList: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
		roles: [Role!]!
	}
	
	type GameMessage {
		sender: String!
		message: String!
		timestamp: String!
	}
	
	type Game implements PermissionControlled{
		_id: ID!
		world: World!
		map: Place
		players: [User!]!
		messages: [GameMessage!]!
		host: User!
		accessControlList: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
		strokes: [Stroke!]!
		models: [PositionedModel!]!
	}
	
	type Stroke{
		_id: String
		path: [PathNode!]!
		type: String!
		size: Int!
		color: String!
		fill: Boolean!
	}
	
	type PathNode{
		x: Float!
		y: Float!
		_id: String
	}
	
	input PathNodeInput{
		x: Float!
		y: Float!
		_id: String
	}
	
	type Model implements PermissionControlled{
		_id: ID!
		name: String!
		depth: Float!
		width: Float!
		height: Float!
		fileName: String!
		fileId: ID!
		accessControlList: [PermissionAssignment!]!
		canWrite: Boolean!
		canAdmin: Boolean!
	}
	
	type PositionedModel {
		_id: ID!
		model: Model!
		x: Float!
		z: Float!
		rotation: Float!
	}
	
`;