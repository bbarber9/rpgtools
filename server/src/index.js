import express from 'express';
import bodyParser from "body-parser";
import gqlServer from './gql-server';
import cookieParser from 'cookie-parser';
import morgan from 'morgan';
import mongoose from 'mongoose';
import {defaultSeeders} from "./default-seeders";
import path from 'path';
import {ImageRouter} from "./routers/image-router";
import {checkConfig, serverNeedsSetup} from './server-needs-setup';
import crypto from 'crypto';
import {ModelRouter} from "./routers/model-router";

const mongodb_host = process.env.MONGODB_HOST || "mongodb";
const mongodb_db_name = process.env.MONGODB_DB_NAME || "rpgtools";

mongoose.connect(`mongodb://${mongodb_host}/${mongodb_db_name}`, {useNewUrlParser: true}).then(
	async () => {
		console.log(`Connected to mongodb at mongodb://${mongodb_host}/${mongodb_db_name}`);
		await defaultSeeders();
		await createServer();
	}
);

mongoose.set('useNewUrlParser', true);
mongoose.set('useFindAndModify', false);
mongoose.set('useCreateIndex', true);
mongoose.set('useUnifiedTopology', true);

const createServer = async () => {
	const server = express();

	await checkConfig();

	server.get('*.js', function (req, res, next) {
		req.url = req.url + '.gz';
		res.set('Content-Encoding', 'gzip');
		res.set('Content-Type', 'text/javascript');
		next();
	});
	server.get('*.css', function (req, res, next) {
		req.url = req.url + '.gz';
		res.set('Content-Encoding', 'gzip');
		res.set('Content-Type', 'text/css');
		next();
	});
	server.get('*favicon.ico', function (req, res, next) {
		req.url = '/favicon.ico';
		next();
	});

	server.use('/images', ImageRouter);
	// server.use(ModelRouter);

	server.get('*', (req, res, next) => {
		const needsSetup = serverNeedsSetup();
		if(needsSetup){
			if (['/api', '/ui/setup', '/app.bundle.js'].includes(req.url)){
				return next();
			}
			return res.redirect(302, '/ui/setup');
		}
		else {
			return next();
		}
	});

	server.get('/ui*', (req, res) => {
		return res.sendFile(path.resolve(__dirname, '../../dist', 'index.html'));
	});


	server.use(express.static('../dist'));
	server.use(express.static('src/static-assets'));

	server.use(bodyParser.json());
	server.use(morgan('tiny'));

	server.use(cookieParser());

	gqlServer.applyMiddleware({app: server, path: '/api'});

	if(!process.env.ACCESS_TOKEN_SECRET){
		console.warn("environment variable ACCESS_TOKEN_SECRET is not set, restarting server will log out all users");
		process.env.ACCESS_TOKEN_SECRET = crypto.randomBytes(2048);
	}

	if(!process.env.REFRESH_TOKEN_SECRET){
		console.log("environment variable REFRESH_TOKEN_SECRET is not set, restarting server will log out all users");
		process.env.REFRESH_TOKEN_SECRET = crypto.randomBytes(2048);
	}

	const port = process.env.SERVER_PORT || 3000;

	server.listen(port, () => {
		console.log(`The server is running and listening at http://localhost:${port}`);
	});

};