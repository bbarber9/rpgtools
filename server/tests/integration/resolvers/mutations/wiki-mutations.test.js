import {User} from "../../../../src/models/user";
import {ANON_USERNAME} from "../../../../src/authentication-helpers";
import {ApolloServer} from "apollo-server-express";
import {typeDefs} from "../../../../src/gql-server-schema";
import {serverResolvers} from "../../../../src/resolvers/server-resolvers";
import {createTestClient} from "apollo-server-testing";
import {CREATE_WIKI, UPDATE_WIKI} from "../../../../../common/src/gql-queries";
import {createWorld} from "../../../../src/resolvers/mutations/world-mutations";
import {Readable} from 'stream';

process.env.TEST_SUITE = 'wiki-mutations-test';

describe('user mutations', () => {
	let currentUser = new User({username: ANON_USERNAME});

	const server = new ApolloServer({
		typeDefs,
		resolvers: serverResolvers,
		context: () => {
			return {
				currentUser: currentUser,
				res: {
					cookie: () => {
					}
				}
			};
		}
	});

	const {mutate} = createTestClient(server);

	describe('with world', () => {

		let world = null;

		beforeEach(async () => {
			const user = await User.findOne({username: 'tester'});
			await user.recalculateAllPermissions();
			world = await createWorld('Earth', false, user);
		});

		test('create wiki no permission', async () => {
			const result = await mutate({mutation: CREATE_WIKI, variables: {name: 'new wiki', folderId: world.rootFolder.toString()}});
			expect(result).toMatchSnapshot();
		});

		test('update wiki no permission', async () => {
			const result = await mutate({mutation: UPDATE_WIKI, variables: {wikiId: world.wikiPage._id.toString(), name: 'new name'}});
			expect(result).toMatchSnapshot();
		});

		describe('with authenticated user', () => {
			beforeEach(async () => {
				currentUser = await User.findOne({username: 'tester'}).populate({path: 'roles', populate: {path: 'permissions'}});
				await currentUser.recalculateAllPermissions();
			});

			test('create wiki', async () => {
				const result = await mutate({mutation: CREATE_WIKI, variables: {name: 'new wiki', folderId: world.rootFolder.toString()}});
				expect(result).toMatchSnapshot({
					data: {
						createWiki: {
							_id: expect.any(String),
							folders: expect.arrayContaining([expect.objectContaining({
								_id: expect.any(String),
								pages: expect.arrayContaining([expect.objectContaining({
									_id: expect.any(String)
								})])
							})])
						}
					}
				});
			});

			test('update wiki just name', async () => {
				const result = await mutate({mutation: UPDATE_WIKI, variables: {wikiId: world.wikiPage._id.toString(), name: 'new name'}});
				expect(result).toMatchSnapshot({
					data: {
						updateWiki: {
							_id: expect.any(String),
							world: {
								_id : expect.any(String)
							}
						}
					}
				});
			});

			test('update wiki just content', async () => {
				const content = {
					createReadStream: () => Readable.from('wiki content'.repeat(1024))
				};
				const result = await mutate({mutation: UPDATE_WIKI, variables: {wikiId: world.wikiPage._id.toString(), content: content}});
				expect(result).toMatchSnapshot({
					data: {
						updateWiki: {
							_id: expect.any(String),
							world: {
								_id : expect.any(String)
							},
							content: expect.any(String)
						}
					}
				});
			});
		});
	});

});