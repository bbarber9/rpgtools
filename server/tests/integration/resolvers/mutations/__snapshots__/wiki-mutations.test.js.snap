// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user mutations with world create wiki no permission 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Authentication is required for this action],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user mutations with world update wiki no permission 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to write to this page],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user mutations with world with authenticated user create wiki 1`] = `
Object {
  "data": Object {
    "createWiki": Object {
      "_id": Any<String>,
      "folders": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "pages": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
            },
          ],
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user mutations with world with authenticated user update wiki just content 1`] = `
Object {
  "data": Object {
    "updateWiki": Object {
      "_id": Any<String>,
      "canWrite": true,
      "content": Any<String>,
      "coverImage": null,
      "name": "Earth",
      "type": "Place",
      "world": Object {
        "_id": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user mutations with world with authenticated user update wiki just name 1`] = `
Object {
  "data": Object {
    "updateWiki": Object {
      "_id": Any<String>,
      "canWrite": true,
      "content": null,
      "coverImage": null,
      "name": "new name",
      "type": "Place",
      "world": Object {
        "_id": Any<String>,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
